_posts/2019-04-05-omakase-typescript.markdown

```diff
@@ -0,0 +1,171 @@
+
+### What do Programming Languages do?
+
+Programming languages are an interesting problem to solve. People read code many, many multiples of times more than
+they write it - so developers create languages which are good at solving particular problems in a terse amount of
+code. Here's an example using JavaScript:
+
+```js
+var name = "Danger"
+console.log("Hello, " + danger)
+```
+
+The first line makes a variable (something you can store something in) and then the seconds outputs text out to the
+console (think DOS, or the terminal) `"Hello, Danger"`. JavaScript is designed to work as a scripting language,
+which means the code starts at the top of the file and then goes through line by line. To provide some contrast,
+here is the [same behavior](https://repl.it/repls/VioletredGlisteningInfo) in Java, which is built with different
+language constraints:
+
+```java
+class Main {
+  public static void main(String[] args) {
+    String name = "Danger";
+    System.out.println("Hello, " + name);
+  }
+}
+```
+
+> Note: if you find the naming of Java and JavaScript confusing (it is) - that's because when JavaScript as being
+> created, Java was looking to be really the next hot language to work with (it did turn out that way for a few
+> decades, but now JavaScript is usually the first language people have heard of.)
+
+Aside from having a lot more lines, the Java version comes with a lot of words that aren't necessarily about
+telling the computer exactly what to do, e.g. `class Main {`, `public static void main(String[] args) {`, `}` and
+`}` again. It also has semi-colons at the end of some line. Java is aimed at building different things, and these
+extras to the code sense within the constraints of building a Java app.
+
+To get to my main point though, there is one standout line I'd like us to compare:
+
+```
+// JavaScript
+var name = "Danger"
+// Java
+String name = "Danger";
+```
+
+Both of these lines set a variable to a be "Danger" as a string (a collection of characters, which in programming
+we call a string (this [5m video](https://www.youtube.com/watch?v=czTWbdwbt7E) is a good primer)) - in JS you write
+`var` which tells the next word after that is the name of the variable. However in Java, the first thing you write
+is `String` instead.
+
+There are many values which can go into a variable, and `var` accepts all of them in JS. In Java though, we have to
+say upfront what can be inside that variable, which is only strings, so:
+
+```js
+// Before in JS
+var name = "Danger"
+// Also OK
+var name = 1
+var name = false
+var name = { first: "Danger", last: "McMaslen" }
+
+// Before in Java
+String name = "Danger";
+// Not OK, the code wouldn't be accepted by Java
+String name = 1;
+
+// This example isn't quite what you'd write, but it shows the difference well
+Map<String, String> person = new HashMap<String, String>();
+person.put("first", "Danger");
+person.put("last", "McMaslen");
+String name = person;
```

Comment:
not sure what's going on here. Might need more comments. Also hash maps and generics are probably going to just confuse people who don't know what a type system is.

Reply:
I really couldn't find some other code that correlates to `var name = { first: "Danger", last: "McMaslen" }` in a reasonable way in Java. But only having two cases (int/bool) felt a bit too few

Reply:
and I don't think you can skip the generics, but I'll see

Reply:
I can do an array instead, from the above paragraph